static/js/latest-news.js

```diff
@@ -0,0 +1,115 @@
+function formatDate(date) {
+  const parsedDate = new Date(date);
+  const monthNames = [
+    "January",
+    "February",
+    "March",
+    "April",
+    "May",
+    "June",
+    "July",
+    "August",
+    "September",
+    "October",
+    "November",
+    "December"
+  ];
+
+  const day = parsedDate.getDate();
+  const monthIndex = parsedDate.getMonth();
+  const year = parsedDate.getFullYear();
+
+  return day + " " + monthNames[monthIndex] + " " + year;
+}
+
+function htmlForLatestArticles(articles) {
+  const articlesTree = document.createDocumentFragment();
+  for (let i = 0; i < articles.length; i++) {
+    const article = articles[i];
+    const div = document.createElement("div");
+    div.classList.add("col-3");
+    const header = document.createElement("h4");
+
+    const link = document.createElement("a");
+    link.href = article.link;
+    link.innerHTML = article.title.rendered;
+
+    const date = document.createElement("p");
+    date.classList.add("u-no-padding--top");
+    const em = document.createElement("em");
+    const time = document.createElement("time");
+    time.setAttribute("pubdate", true);
+    time.datetime = article.date;
+    time.innerHTML = formatDate(article.date);
+
+    date.appendChild(time);
+    date.appendChild(em);
+    header.appendChild(link);
+
+    div.appendChild(header);
+    div.appendChild(date);
+
+    articlesTree.appendChild(div);
+  }
+  return articlesTree;
+}
+
+function htmlForLatestPinnedArticle(article) {
+  const articlesTree = document.createDocumentFragment();
+
+  const heading3 = document.createElement("h3");
+  heading3.innerHTML = "Spotlight";
+  const innerDiv = document.createElement("div");
+  const heading4 = document.createElement("h4");
+  const link = document.createElement("a");
+  link.href = article.title;
+  link.innerHTML = article.title.rendered;
+
+  const paragraph = document.createElement("p");
+  paragraph.classList.add("u-no-padding--top");
+  const em = document.createElement("em");
+  const time = document.createElement("time");
+  time.setAttribute("pubdate", true);
+  time.datetime = article.date;
+  time.innerHTML = formatDate(article.date);
+
+  em.appendChild(time);
+  paragraph.appendChild(em);
+  heading4.appendChild(link);
+
+  innerDiv.appendChild(heading4);
+  innerDiv.appendChild(paragraph);
+
+  articlesTree.appendChild(heading3);
+  articlesTree.appendChild(innerDiv);
+
+  return articlesTree;
+}
+
+function reqListener() {
+  const data = JSON.parse(this.responseText);
+  const latest = data.latest_articles[0];
+  const latestPinned = latest[0];
+
+  const containerForLatestArticles = document.getElementById("latest-articles");
+  const html = htmlForLatestArticles(latest);
+  containerForLatestArticles.appendChild(html);
+
+  if (latestPinned) {
+    const containerForLatestNews = document.getElementById(
+      "latest-news-container"
```

Comment:
You have ids of the elements hardcoded in JS.
It's not wrong for something simple, but it may be hard to look for it when something changes in the template.
It would be nice at least to have these extracted to some variables (so they can be found in one place in top of the module for example). Or ideally they should be passed in as a parameters from the template itself.

Reply:
Hm, what do you think of exposing a function on the window?
That could then be called with the correct IDs from the template.

Reply:
Yes personally I like the pattern of all JavaScript files only ever defining functions, and then calling those functions from inside the page to initialise the actual action, and pass through any settings:

``` html
<script src="make-element-awesome.js"></script>
<script>makeElementAwesome(elementId="awesomeable-element");</script>
```

I have a feeling others might disagree with initialising JS functions directly in the markup, but at least making any settings into arguments is a nice pattern that I think you should consider.

Reply:
@nottrobin We do it in snapcraft. It has some downside of requiring some global namespace for those 'init functions' and requires specific webpack config to allow that.

I looked back then at what webpack recommends and it appears that ideally you shouldn't pass anything to modules (they should be self contained). So in such case one module should export class/function that can be parametrised. And there should be a small module that call that with some options - like we do here with importing global nav and running it with width param.

Other things they recommend is making a script that reads data from the HTML, not HTML pass data via calling a function.
Not sure if I recall any docs mentioning that, but I found an issue where it is discussed:
https://github.com/webpack/webpack/issues/2683

Reply:
So, I have added a namespaced object on the window called `canonicalWebteam`. The module then registers itself on this object, if it has not been defined before, thus making it accessible in the template to pass in DOM anchors.

I have also updated this to only needing one root anchor, the other element is then selected from inside that container context with `querySelector` instead of `getElementById`.

Is this too complex/overengineered? Or is this sth we should consider doing? @bartaz @nottrobin

Reply:
@b-m-f Did you commit it here?

If we want (and need) to expose modules as functions to pass some parameters to them I'd rather do it on build level than having some own solution of adding them to global object.
In snapcraft we are using `expose-loader` plugin for that, but it may be possible in webpack directly via `output.library` configuration (I just haven't know about it before https://webpack.js.org/configuration/output).

The question how to deal with passing params to modules and should we export them to global functions is a dev catch-up material maybe?
I find it useful, understandable and explicit (just adding the script to page doesn't mean it runs, you need to call it), but it's also not seen as a good practice from webpack (https://github.com/webpack/webpack/issues/2683#issuecomment-228181205), so I'm not sure if that's the solution we want to move forward with.

Reply:
@bartaz forgot to push :sweat_smile: . Its up now.

Yup, might be a good dev-catch up topic. 

So how should we procede here @nottrobin @bartaz ?

Go with the proposed solution, or archive the latest commit and go with the previous (script has knowledge about HTML version).